### DISCUSSION TITLE
How to make subtrees with specific number of variables?

### DISCUSSION THREAD

**Discussion by DenisSvirin** [MilesCranmer/PySR#467] [Q&A] [answered]
I am using custom loss to divide tree into several subtrees.
```
tree.degree != 2 && return L(Inf)

 P = tree.l
 Q = tree.r 
```

**But i need P to have 1 variable [P(x)] and Q to have 3 variables
[Q(x1, x2, x3)]. **

I guess, i need to use training set with 3 features, but i cant figure out how to restrain P to have only 1.

**Comment by MilesCranmer** [comment, marked as answer]
Maybe this helps? https://github.com/MilesCranmer/PySR/discussions/465

**Comment by DmitriiKumancev** [comment]
Assuming that your tree represents a decision or regression model, and P(x) and Q(x1, x2, x3) are the predictions of the model for different subtrees, you can structure your custom loss function accordingly.

Here's a simplified example in Python using TensorFlow:
`import tensorflow as tf

def custom_loss(y_true, y_pred):
    # Assuming y_true and y_pred are the true values and predicted values, respectively

    # Split the predicted values into P and Q
    P = y_pred[:, 0]  # Assuming the first column corresponds to P
    Q = y_pred[:, 1:]  # Assuming the remaining columns correspond to Q's variables

    # Define your constraints on P and Q in the loss function
    loss_P = tf.reduce_mean(tf.square(P - y_true))  # Example constraint on P

    # Example constraint on Q's variables
    constraint_Q = tf.reduce_mean(tf.square(Q[:, 0] + Q[:, 1] + Q[:, 2] - y_true))

    # Combine the individual losses with weights or other constraints as needed
    total_loss = loss_P + constraint_Q

    return total_loss

# Example usage:
# model.compile(optimizer='adam', loss=custom_loss)
# model.fit(train_features, train_labels)
`

In this example, the model predicts both P and Q, and the loss function includes constraints on P and Q. You can customize the constraints based on your specific requirements. Ensure that your model architecture outputs the appropriate number of values for P and Q.

It's important to adapt this example to your specific model architecture, requirements, and constraints. The goal is to design a loss function that guides the training process to achieve the desired behavior for P and Q in your decision tree.

**Reply by MilesCranmer** [reply]
Sorry, did you write this in ChatGPT? It sounds like the exact same tone and it’s also completely unrelated to the question... Maybe if you build a PySR GPT with https://github.com/BuilderIO/gpt-crawler it would do it better. 

But would be good to state if you are using GPT so people can double check code 😓
